haskell/src/Lentille/GraphQL.hs

```diff
@@ -110,17 +136,55 @@ data RateLimit = RateLimit {used :: Int, remaining :: Int, resetAt :: UTCTime}
 instance From RateLimit Text where
   from RateLimit {..} = "remains:" <> show remaining <> ", reset at: " <> show resetAt
 
+type RetryCheck m a = (Either (FetchError a) a, [ReqLog]) -> m Bool
+
+-- | wrapper around fetchWithLog than can optionaly handle fetch retries
+-- based on the returned data inspection via a provided function (see RetryCheck).
+-- In case of retry the depth parameter of mkArgs is decreased (see adaptDepth)
+doRequest ::
+  forall a.
+  forall m.
+  (MonadGraphQLE m, Fetch a, FromJSON a) =>
+  GraphClient ->
+  (Maybe Int -> Maybe Text -> Args a) ->
+  Maybe (RetryCheck m a) ->
+  Maybe Int ->
+  Maybe PageInfo ->
```

Comment:
nit: when parameters are not likely to be pre-applied (e.g. curry), then they can be named using record field, e.g.:

```haskell
data RequestParams m a = RequestParams {
    fpMkArgs :: Maybe Int -> Maybe Text -> Args a
  , fpRetryCheck :: Maybe (RetryCheck m a)
  , ...
  }
```

Then the function would be: `doRequest :: GraphClient -> RequestParams m a -> m a`

And we can provide a `defaultRequestsParams = RequestParams Nothing Nothing Nothing`

Reply:
Thanks you for the tips. As of now doRequest is only used by streamFetch so it does not make so much sense to do it there. However streamFetch might be a better candidate.

Reply:
This has been implemented in a81b8b01dfbea551926f5734819dbd73cde340a6